airbyte-workers/src/main/java/io/airbyte/workers/internal/EmptyAirbyteSource.java

```diff
@@ -61,4 +120,109 @@ public void cancel() throws Exception {
     // no op.
   }
 
+  private Optional<AirbyteMessage> emitStreamState() {
+    // Per stream, it will emit one message per stream being reset
+    if (!streamDescriptors.isEmpty()) {
+      try {
+        final StreamDescriptor streamDescriptor = streamDescriptors.poll();
+        return Optional.of(getNullStreamStateMessage(streamDescriptor));
+      } catch (final EmptyStackException e) {
+        return Optional.empty();
+      }
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  private Optional<AirbyteMessage> emitGlobalState() {
+    if (hasEmittedState.get()) {
+      return Optional.empty();
+    } else {
+      hasEmittedState.compareAndSet(false, true);
+      return Optional.of(getNullGlobalMessage(streamDescriptors, stateWrapper.get().getGlobal()));
+    }
+  }
+
+  private Optional<AirbyteMessage> emitLegacyState() {
+    if (!hasEmittedState.get()) {
+      hasEmittedState.compareAndSet(false, true);
+      return Optional.of(new AirbyteMessage().withType(Type.STATE)
+          .withState(new AirbyteStateMessage().withStateType(AirbyteStateType.LEGACY).withData(Jsons.emptyObject())));
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  private boolean configHasFullCatalog(final WorkerSourceConfig sourceConfig) {
+    final Set<StreamDescriptor> catalogStreamDescriptors = sourceConfig.getCatalog().getStreams().stream().map(
+            configuredAirbyteStream -> new StreamDescriptor()
+                .withName(configuredAirbyteStream.getStream().getName())
+                .withNamespace(configuredAirbyteStream.getStream().getNamespace()))
+        .collect(Collectors.toSet());
+    final Set<StreamDescriptor> configStreamDescriptors = new HashSet<>(streamDescriptors);
+
+    return catalogStreamDescriptors.equals(configStreamDescriptors);
+  }
+
+  private AirbyteMessage getNullStreamStateMessage(final StreamDescriptor configStreamDescriptor) {
```

Comment:
nit: rename `configStreamDescriptor` -> `streamToReset`

Reply:
can it be `static`?

Reply:
It can be static but I don't see the benefits of it since the method isn't call outside this class or by a public static method. If we want to make them public I think that it should be extracted in a dedicated class and be static.